src/main/java/io/kontur/disasterninja/notifications/email/EmailNotificationService.java

```diff
@@ -33,6 +35,12 @@ public class EmailNotificationService extends NotificationService {
     private final LayersApiService layersApiService;
     private final GeometryTransformer geometryTransformer;
 
+    private final ConcurrentHashMap<UUID, Long> notificationTimestamps = new ConcurrentHashMap<>();
```

Comment:
ConcurrentHashMap is used because the implementation of Guava and Caffeine cache has bugs, which cause race conditions and can lead to incorrect behavior. For our simple case, the map with a manual cleaning function should be enough.

Reply:
I suggest putting your comment here into a comment in the code. 😉

Reply:
Please open a ticket if you believe there are bugs. Thanks!